js/runner.go

```diff
@@ -771,24 +776,35 @@ func (u *VU) runFn(
 		u.state.Tags.Set("iter", strconv.FormatInt(u.state.Iteration, 10))
 	}
 
-	defer func() {
-		if r := recover(); r != nil {
-			gojaStack := u.Runtime.CaptureCallStack(20, nil)
-			err = fmt.Errorf("a panic occurred in VU code but was caught: %s", r)
-			// TODO figure out how to use PanicLevel without panicing .. this might require changing
-			// the logger we use see
-			// https://github.com/sirupsen/logrus/issues/1028
-			// https://github.com/sirupsen/logrus/issues/993
-			b := new(bytes.Buffer)
-			for _, s := range gojaStack {
-				s.Write(b)
-			}
-			u.state.Logger.Log(logrus.ErrorLevel, "panic: ", r, "\n", string(debug.Stack()), "\nGoja stack:\n", b.String())
-		}
-	}()
-
 	startTime := time.Now()
-	v, err = fn(goja.Undefined(), args...) // Actually run the JS script
+
+	if u.moduleVUImpl.eventLoop == nil {
+		u.moduleVUImpl.eventLoop = newEventLoop(u.moduleVUImpl)
+	}
+	err = u.moduleVUImpl.eventLoop.start(func() (err error) {
+		// here the returned value purposefully shadows the external one as they can different
+		defer func() {
+			if r := recover(); r != nil {
+				gojaStack := u.Runtime.CaptureCallStack(20, nil)
+				err = fmt.Errorf("a panic occurred in VU code but was caught: %s", r)
+				// TODO figure out how to use PanicLevel without panicing .. this might require changing
+				// the logger we use see
+				// https://github.com/sirupsen/logrus/issues/1028
+				// https://github.com/sirupsen/logrus/issues/993
+				b := new(bytes.Buffer)
+				for _, s := range gojaStack {
+					s.Write(b)
+				}
+				u.state.Logger.Log(logrus.ErrorLevel, "panic: ", r, "\n", string(debug.Stack()), "\nGoja stack:\n", b.String())
+			}
+		}()
+		v, err = fn(goja.Undefined(), args...) // Actually run the JS script
+		return err
+	})
+	if isDefault && err != nil {
```

Comment:
```suggestion
	if err != nil && isDefault {
```
It's more a personal preference, if there is a specific motivation then I'm not getting it.